src/bot.rs

```diff
@@ -59,151 +60,283 @@ impl Bot {
-    pub fn fetch_formdata<'a, T>(
+    pub fn fetch_formdata<T>(
         &self,
-        func: &str,
+        func: String,
         msg: Value,
-        mut file: T,
+        file: T,
         kind: &str,
         file_name: &str,
-    ) -> impl Future<Item = String, Error = Error> + 'a
+    ) -> impl Future<Item = String, Error = Error>
     where
         T: io::Read,
     {
-        debug!("Send formdata: {}", msg.to_string());
-        let mut content = Vec::new();
+        fetch_formdata(
+            self.session.clone(),
+            self.key.clone(),
+            func,
+            msg,
+            file,
+            String::from(kind),
+            String::from(file_name),
+        )
+    }
+}
+
+/// Creates a new request and adds a JSON message to it. The returned Future contains a the
+/// reply as a string.  This method should be used if no file is added because a JSON msg is
+/// always compacter than a formdata one.
+pub fn fetch_json(
+    session: Session,
+    key: String,
+    func: String,
+    msg: String,
+) -> impl Future<Item = String, Error = Error> {
+    debug!("Send JSON: {}", msg);
+
+    result(Ok(List::new()))
+        .and_then(move |mut header| {
+            header
+                .append("Content-Type: application/json")
+                .map(|_| header)
+                .map_err(|e| {
+                    error!("Error: {}", e);
+
+                    Error::Unknown
+                })
+        })
+        .and_then(move |header| {
+            let mut a = Easy::new();
+
+            a.http_headers(header).map(|_| a).map_err(|e| {
+                error!("Error: {}", e);
+
+                Error::Unknown
+            })
+        })
+        .and_then(move |mut a| {
+            a.post_fields_copy(msg.as_bytes()).map(|_| a).map_err(|e| {
+                error!("Error: {}", e);
 
-        let mut a = Easy::new();
-        let mut form = Form::new();
+                Error::Unknown
+            })
+        })
+        .and_then(move |mut a| {
+            a.post(true).map(|_| a).map_err(|e| {
+                error!("Error: {}", e);
+
+                Error::Unknown
+            })
+        })
+        .and_then(move |a| _fetch(session, key, func, a))
+}
 
-        // try to read the byte content to a vector
-        let _ = file.read_to_end(&mut content).unwrap();
+/// Creates a new request with some byte content (e.g. a file). The method properties have to be
+/// in the formdata setup and cannot be sent as JSON.
+pub fn fetch_formdata<T>(
+    session: Session,
+    key: String,
+    func: String,
+    msg: Value,
+    mut file: T,
+    kind: String,
+    file_name: String,
+) -> impl Future<Item = String, Error = Error>
+where
+    T: io::Read,
+{
+    debug!("Send formdata: {}", msg.to_string());
+
+    result(Ok(Vec::new()))
+        .and_then(move |mut content| {
+            // try to read the byte content to a vector
+            file.read_to_end(&mut content).map(|_| content).map_err(
```

Comment:
Here is the same scenario as above, perhaps it would be great if all results are collected into a single and then propagated to the future. This would avoid multiple map and map_err's.